### PR TITLE
fix: propagate generics from FastifyRegister to plugin type

### LINKED ISSUE #4032: FastifyRegister type does not propagate type provider generic to plugin type
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.x.x

### Plugin version

_No response_

### Node.js version

16.x

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

11.6

### Description

The `FastifyRegister` type (of `fastify.register` function) has this interface:

```ts
export interface FastifyRegister<T = void> {
  <Options extends FastifyPluginOptions>(
    plugin: FastifyPluginCallback<Options>,
    opts?: FastifyRegisterOptions<Options>
  ): T;
  <Options extends FastifyPluginOptions>(
    plugin: FastifyPluginAsync<Options>,
    opts?: FastifyRegisterOptions<Options>
  ): T;
  <Options extends FastifyPluginOptions>(
    plugin: FastifyPluginCallback<Options> | FastifyPluginAsync<Options> | Promise<{ default: FastifyPluginCallback<Options> }> | Promise<{ default: FastifyPluginAsync<Options> }>,
    opts?: FastifyRegisterOptions<Options>
  ): T;
}
```

This passes through the `Options` to the `plugin` but not the other arguments (`Server` and `TypeProvider`) which results in an error like this:

```
No overload matches this call.
  Overload 1 of 3, '(plugin: FastifyPluginCallback<Record<never, never>, http.Server, FastifyTypeProviderDefault>, opts?: FastifyRegisterOptions<Record<never, never>> | undefined): FastifyInstance<...> & PromiseLike<...>', gave the following error.
    Argument of type 'FastifyPluginAsync<Record<never, never>, http.Server, TypeBoxTypeProvider>' is not assignable to parameter of type 'FastifyPluginCallback<Record<never, never>, http.Server, FastifyTypeProviderDefault>'.
      Types of parameters 'instance' and 'instance' are incompatible.
        Type 'FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse, FastifyLoggerInstance, FastifyTypeProviderDefault>' is not assignable to type 'FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse, FastifyLoggerInstance, TypeBoxTypeProvider>'.
          The types returned by 'after()' are incompatible between these types.
            Type 'FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse, FastifyLoggerInstance, FastifyTypeProviderDefault> & PromiseLike<undefined>' is not assignable to type 'FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse, FastifyLoggerInstance, TypeBoxTypeProvider> & PromiseLike<undefined>'.
              Type 'FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse, FastifyLoggerInstance, FastifyTypeProviderDefault> & PromiseLike<undefined>' is not assignable to type 'FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse, FastifyLoggerInstance, TypeBoxTypeProvider>'.
                Types of property 'register' are incompatible.
                  Type 'FastifyRegister<FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse, FastifyLoggerInstance, FastifyTypeProviderDefault> & PromiseLike<undefined>>' is not assignable to type 'FastifyRegister<FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse, FastifyLoggerInstance, TypeBoxTypeProvider> & PromiseLike<undefined>>'.
                    Type 'FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse, FastifyLoggerInstance, FastifyTypeProviderDefault> & PromiseLike<undefined>' is not assignable to type 'FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse, FastifyLoggerInstance, TypeBoxTypeProvider> & PromiseLike<undefined>'.
                      Type 'FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse, FastifyLoggerInstance, FastifyTypeProviderDefault> & PromiseLike<undefined>' is not assignable to type 'FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse, FastifyLoggerInstance, TypeBoxTypeProvider>'.
                        The types of 'addSchema(...).get' are incompatible between these types.
                          Type 'RouteShorthandMethod<http.Server, http.IncomingMessage, http.ServerResponse, FastifyTypeProviderDefault>' is not assignable to type 'RouteShorthandMethod<http.Server, http.IncomingMessage, http.ServerResponse, TypeBoxTypeProvider>'.
                            Type 'FastifyTypeProviderDefault' is not assignable to type 'TypeBoxTypeProvider'.
  Overload 2 of 3, '(plugin: FastifyPluginAsync<Record<never, never>, http.Server, FastifyTypeProviderDefault>, opts?: FastifyRegisterOptions<Record<never, never>> | undefined): FastifyInstance<...> & PromiseLike<...>', gave the following error.
    Argument of type 'FastifyPluginAsync<Record<never, never>, http.Server, TypeBoxTypeProvider>' is not assignable to parameter of type 'FastifyPluginAsync<Record<never, never>, http.Server, FastifyTypeProviderDefault>'.
      Types of parameters 'instance' and 'instance' are incompatible.
        Type 'FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse, FastifyLoggerInstance, FastifyTypeProviderDefault>' is not assignable to type 'FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse, FastifyLoggerInstance, TypeBoxTypeProvider>'.
  Overload 3 of 3, '(plugin: FastifyPluginCallback<Record<never, never>, http.Server, FastifyTypeProviderDefault> | FastifyPluginAsync<Record<never, never>, http.Server, FastifyTypeProviderDefault> | Promise<...> | Promise<...>, opts?: FastifyRegisterOptions<...> | undefined): FastifyInstance<...> & PromiseLike<...>', gave the following error.
    Argument of type 'FastifyPluginAsync<Record<never, never>, http.Server, TypeBoxTypeProvider>' is not assignable to parameter of type 'FastifyPluginCallback<Record<never, never>, http.Server, FastifyTypeProviderDefault> | FastifyPluginAsync<Record<never, never>, http.Server, FastifyTypeProviderDefault> | Promise<...> | Promise<...>'.
      Type 'FastifyPluginAsync<Record<never, never>, http.Server, TypeBoxTypeProvider>' is not assignable to type 'FastifyPluginAsync<Record<never, never>, http.Server, FastifyTypeProviderDefault>'.
```

### Steps to Reproduce

See https://github.com/stefee/fastify-issue-repro-4032/blob/main/index.ts

### Expected Behavior

No type error.

### PR DESCRIPTION
**Pull request by stefee**
Closes #4032

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by RafaelGSS — COMMENTED** [review]
Can you include a tsd test as well?

**Comment by stefee** [comment]
@RafaelGSS Yep, will do when I have time.

**Comment by stefee** [comment]
Fixed commit message.

**Comment by stefee** [comment]
Changing back to a draft - there's a failing test.

**Comment by stefee** [comment]
The failing test is in `hook.test-d.ts` 

https://github.com/stefee/fastify/blob/e68bd8c42b5cbcabcb4cfee860e1e8d87a730b9c/test/types/hooks.test-d.ts#L217-L236

```
test/types/hooks.test-d.ts:235:19
  ✖  235:19  No overload matches this call.
  The last overload gave the following error.
    Argument of type "preHandler" is not assignable to parameter of type "onClose".

  1 error
```

The problem is to do with passing anonymous (and therefore implicitly typed) plugin into `.register(`. It seems that the type of `instance` that gets passed to the plugin has a concrete type which has changed as a result of the change in this PR.

**Comment by stefee** [comment]
The above issue was resolved by providing default values inferred from the instance on which `.register` is called.

**Comment by stefee** [comment]
Moved tests into the relevant file.

**Review by RafaelGSS — APPROVED** [review]
LGTM.

**Comment by RafaelGSS** [comment]
cc: @sinclairzx81

File: types/register.d.ts [hunk 1]
```diff
@@ -15,17 +17,17 @@ export type FastifyRegisterOptions<Options> = (RegisterOptions & Options) | ((in
...
-  <Options extends FastifyPluginOptions>(
-    plugin: FastifyPluginAsync<Options>,
+  <Options extends FastifyPluginOptions, Server extends RawServerBase = RawServer, TypeProvider extends FastifyTypeProvider = TypeProviderDefault>(
+    plugin: FastifyPluginAsync<Options, Server, TypeProvider>,
     opts?: FastifyRegisterOptions<Options>
   ): T;
-  <Options extends FastifyPluginOptions>(
-    plugin: FastifyPluginCallback<Options> | FastifyPluginAsync<Options> | Promise<{ default: FastifyPluginCallback<Options> }> | Promise<{ default: FastifyPluginAsync<Options> }>,
+  <Options extends FastifyPluginOptions, Server extends RawServerBase = RawServer, TypeProvider extends FastifyTypeProvider = TypeProviderDefault>(
+    plugin: FastifyPluginCallback<Options, Server, TypeProvider> | FastifyPluginAsync<Options, RawServer, TypeProvider> | Promise<{ default: FastifyPluginCallback<Options, RawServer, TypeProvider> }> | Promise<{ default: FastifyPluginAsync<Options, RawServer, TypeProvider> }>,
```
Comment (stefee) [inline types/register.d.ts:30]: I just noticed this error here - I think it doesn't make any difference in practice but it's not consistent with the other cases:

```suggestion
    plugin: FastifyPluginCallback<Options, Server, TypeProvider> | FastifyPluginAsync<Options, Server, TypeProvider> | Promise<{ default: FastifyPluginCallback<Options, Server, TypeProvider> }> | Promise<{ default: FastifyPluginAsync<Options, Server, TypeProvider> }>,
```

**Comment by mcollina** [comment]
@climba03003 @fox1t can you take a look?

**Review by climba03003 — COMMENTED** [review]
I am only afraid will it affect the non-default generic.
For example, people using `http` or `https`.

**Comment by mcollina** [comment]
@sinclairzx81 do you approve this?

**Comment by sinclairzx81** [comment]
@mcollina Hi, this looks ok to me. As suggested by @climba03003, it may be good to put some TS tests up around the `http2` and `https` server types (as these propagate), This would just rule out any issues with varying server types prior to merge if there are concerns there.  

But other than that, this looks great :) @stefee Nice work!

**Comment by stefee** [comment]
@climba03003 @sinclairzx81 @mcollina Which type signature is expected for a **https** server instance?

default:

```
FastifyInstance
```

http2:

```
FastifyInstance<Http2Server, Http2ServerRequest, Http2ServerResponse>
```

with type provider:
```
FastifyInstance<Server, IncomingMessage, ServerResponse, FastifyLoggerInstance, TestTypeProvider>
```

https: ?

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by sinclairzx81** [comment]
> Which type signature is expected for a https server instance?

@stefee The following generic arguments should match each of the varying server types.

```typescript
import * as http from 'http'
import * as https from 'https'
import * as http2 from 'http2'

...

// With Http
const serverWithHttp = fastify({ })
type ServerWithHttp = FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse>

...

// With Https
const serverWithHttps = fastify({ https: { } })
type ServerWithHttps = FastifyInstance<https.Server, http.IncomingMessage, http.ServerResponse>

...

// With Http2
const serverWithHttp2 = fastify({ http2: true })
type ServerWithHttp2 = FastifyInstance<http2.Http2Server, http2.Http2ServerRequest, http2.Http2ServerResponse>
...
```
Have tested the HTTPS variant locally, and all looks good to me :)

Good work!

**Comment by fox1t** [comment]
LGTM! Nice job!

**Pull request merged**